geometric_kernels/kernels/product.py

```diff
@@ -0,0 +1,62 @@
+"""
+Product of kernels
+"""
+from typing import List, Tuple
+
+import lab as B
+
+from geometric_kernels.spaces import Space
+from geometric_kernels.kernels import BaseGeometricKernel
+
+
+class ProductGeometricKernel(BaseGeometricKernel):
+    """
+    Basic implementation of product kernels.
+    TODO: Support combined RFF/KL expansion methods
+    """
+
+    def __init__(self, *kernels):
+        self.kernels = kernels
+
+        dimensions = [kernel.space.dimension for kernel in self.kernels]
+        self.dimension_indices = []
+        i = 0
+        inds = B.linspace(0, sum(dimensions) - 1, sum(dimensions)).astype(int)
+        for dim in dimensions:
+            self.dimension_indices.append(inds[i : i + dim])
+            i += dim
+
+    @property
+    def space(self) -> List[Space]:
+        return [kernel.space for kernel in self.kernels]
+
+    def init_params_and_state(self) -> Tuple(List[dict], List[dict]):
+        params_and_state = [kernel.init_params_and_state() for kernel in self.kernels]
+
+        return [p[0] for p in params_and_state], [s[1] for s in params_and_state]
+
+    def K(self, params, state, X, X2=None, **kwargs) -> B.Numeric:
+        if X2 is None:
+            X2 = X
+
+        Xs = [B.take(X, inds, axis=-1) for inds in self.dimension_indices]
+        X2s = [B.take(X2, inds, axis=-1) for inds in self.dimension_indices]
+
+        return B.stack(
+            *[
+                kernel.K(X, X2, p, s)
+                for kernel, X, X2, p, s in zip(self.kernels, Xs, X2s, params, state)
+            ],
+            axis=-1,
+        ).prod(dim=-1)
+
+    def K_diag(self, params, state, X):
+        Xs = [B.take(X, inds, axis=-1) for inds in self.dimension_indices]
+
+        return B.stack(
+            *[
+                kernel.K_diag(X, p, s)
```

Comment:
I'm surprised by the order of the input arguments. Doesn't seem to match the interface?
https://github.com/GPflow/GeometricKernels/blob/76254b0e36fd41fca6656f63356924db4f16b2b4/geometric_kernels/kernels/base.py#L49